ts.1
.\" ts.1 man page for timestamp
.\" http://math.missouristate.edu/~erik/software.php?id=95
.\"
.\" timestamp - time stamp pipe
.\" 
.\" Copyright (C) 2003-2010 Erik Greenwald <erik@smluc.org> All Rights Reserved.
.\" 
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions are
.\" met:
.\" 
.\" 1. Redistributions of source code must retain the above copyright notice,
.\"    this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. The name of the author may not be used to endorse or promote products
.\"    derived from this software without specific prior written permission.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\" IN  NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
.\" NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
.\" THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.\" $Id: ts.1,v 1.11 2010/01/06 01:06:55 erik Exp $
.\"
.Dd February 14, 2003
.Dt TS 1
.Os
.\"===========================================================================
.Sh NAME
.Nm ts
.Nd timestamp
.\"===========================================================================
.Sh SYNOPSIS
.Nm
.Op Fl uvh
.Op Cm + Ns Ar format
.\"===========================================================================
.Sh DESCRIPTION
Read stdin, timestamp lines, output to stdout. The timestamp format can be set
in the same fashion as 'date' output, and can be set for timezone using the
.Cm TZ
variable, or UTC using the
.Op Fl u
flag. The timestamp for the line is set when the first character of that line
is recieved.
.Pp
The
.Op Cm + Ns Ar format
string is the same syntax as used in the
.Xr date 1
program and the
.Xr srftime 3
library call. If no format string is defined, 
.Nm
will default to
.Tp
.Ns '+%Y%m%d%M%H%S: '
to provide a sort and cut friendly time stamp.
.\"===========================================================================
.Sh OPTIONS
.Bl -tag -width -indent
.It Fl u
Use UTC as the timezone instead of the TZ environment variable or the
/etc/localtime file.
.It Fl h
Show the help screen.
.It Fl v
Show the version.
.El
.\"===========================================================================
.Sh ENVIRONMENT
.Ev TZ Set the time zone to use. -u supercedes this.
.\"===========================================================================
.Sh EXAMPLES
To, say, see what time each of the stages of a FreeBSD buildworld began
.Dl $ make buildworld | grep '^>' | ts
Building a transaction log of events might be something similar to
.Dl $ cat | ts '+%h %d, %Y  %H:%M:%S> ' | tee thing.log
.\"===========================================================================
.Sh DIAGNOSTICS
.Ex -std
.\"===========================================================================
.Sh SEE ALSO
.Xr date 1 ,
.Xr strftime 3
.\"===========================================================================
.Sh AUTHOR
.An "Erik Greenwald" Aq erik@smluc.org
.\"===========================================================================
.Sh BUGS
.Nm
has no known bugs. Please report any found to
.Aq erik@smluc.org


ChangeLog
20071025:
	* splint and flawfinder fixes.

20050318:
	* Version 1.1 released.
	* Minor cleanup.

20040303:
	* Split in to multiple source files.
	* Doxygen shtuff.

20040116:
	* Auto-generated RPM spec file.

20040114:
	* Use of config.h

20030215 (Feb 15, 2003):
	* Version 1.0 released.
	


ts.c

/****************************************************************************\
 * timestamp - time stamp pipe                                               *
 *                                                                           *
 * Copyright (C) 2003-2010 Erik Greenwald <erik@smluc.org> All Rights        *
 * Reserved.                                                                 *
 *                                                                           *
 * Redistribution and use in source and binary forms, with or without        *
 * modification, are permitted provided that the following conditions are    *
 * met:                                                                      *
 *                                                                           *
 * 1. Redistributions of source code must retain the above copyright notice, *
 *    this list of conditions and the following disclaimer.                  *
 * 2. Redistributions in binary form must reproduce the above copyright      *
 *    notice, this list of conditions and the following disclaimer in the    *
 *    documentation and/or other materials provided with the distribution.   *
 * 3. The name of the author may not be used to endorse or promote products  *
 *    derived from this software without specific prior written permission.  *
 *                                                                           *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR      * 
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES *
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.   *
 * IN  NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,         *
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT  *
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, *
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY     *
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT       *
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF  *
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.         *
 *                                                                           *
 * $Id: ts.c,v 1.22 2010/01/06 01:06:55 erik Exp $
 *                                                                           *
 \***************************************************************************/


#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#include <stdio.h>  /* fprintf, stderr, EXIT_* */
#include <stdlib.h> /* free */

#include "opt.h"    /* parse_opts */
#include "stamp.h"  /* stamp */

/**
 * Entry point for timestamp.
 * @param argc Number of arguments.
 * @param argv Argument vector.
 * @return Exit status.
 */
int
main(int argc, char **argv)
{
	const char *format = parse_opts(argc, argv);

	if (format == NULL) {
		fprintf(stderr, "Unable to parse format\n");
		return EXIT_FAILURE;
	}

	if (stamp(format) > 0) {
		fprintf(stderr, "Error timestamping\n");
		free((void *)format);
		return EXIT_FAILURE;
	}

	free((void *)format);

	return EXIT_SUCCESS;
}
